doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -77,22 +77,24 @@ Traditional databases track cardinality data and statistics about various column
 
 ### Symbolic costs
 
-A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In priority order:
 
-| Dimension          | Input features                                               | Outputs                         | Output type |
-| ------------------ | ------------------------------------------------------------ | ------------------------------- | ----------- |
-| Hydration time     | Source cardinality                                           | Time estimate                   |             |
-| Latency per update | Source cardinality, expected per-update delta on each source | Time estimate (per source)      |             |
-| Storage            | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates |             |
-| Scalability        | Distribution of keys (e.g., cardinality of distinct keys)    | Data parallelism estimate       |             |
+| Dimension          | Input features                                               | Outputs                         | Output type                         |
+| ------------------ | ------------------------------------------------------------ | ------------------------------- | ----------------------------------- |
+| Memory             | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates | (Vec<Arrangement>, Formula<Source>) |
+| Hydration time     | Source cardinality                                           | Time estimate                   | Formula<Source>                     |
```

Comment:
nit: Intuitively, I think just saying "cardinality" and "change rate" in the second column (that is, for each input) makes more sense here than restricting those only to sources.

I would imagine that we would end up with following dimension types:

* available arrangements `A` (*)
* predicate selectivity `S` (*)
* cardinality `C`
* change rate `ΔC`
* memory `M`
* hydration time `H`
* per-update latency `L`

Where types marked with (*) are statically available and provided by the environment (in other words, they cannot be inferred), and from the rest each can be inferred recursively from the following input dimensions:

* `C` depends on input features `{ C, S }`
* `ΔC` depends on input features `{ ΔC, S }` (with some caveats)
* `M` depends on input features `{ C, sum(M) }`
* `H) depends on input features `{ C, sum(H) }`
* `L` depends on input features `{ ΔC, L, max(L) }`

Where `sum(-)` and `max(-)` represents a feature derived by taking the corresponding aggregate along the each path.